Configure rustfmt

raskchanky opened the pull request:
It turns out that `rustfmt` supports a configuration file, `.rustfmt.toml`, which can be used to express preferences about how `rustfmt` carries out its duties.  `rustfmt` has both stable and unstable options.  Sadly, unstable options are only available with nightly toolchains, so they are unavailable to us.

This PR configures `rustfmt` in an opinionated way, and reformats this repo according to those rules.  I've listed every stable configuration option in the file, and many of the values there are default values, but I felt there was value in laying everything out explicitly.

If you disagree with any of the options I've chosen, please speak up, and let's have that conversation.  
After this PR merges, I'll apply the same `rustfmt` options to the `core` repo as well.

Closes https://github.com/habitat-sh/habitat/issues/6130

![](https://media.giphy.com/media/1lnpan4G4yGg8/giphy.gif)

thesentinels commented:
Thanks for the pull request! Here is what will happen next:
 1. Your PR will be reviewed by the maintainers
 2. If everything looks good, one of them will `approve` it, and your PR will be merged.

Thank you for contributing!

raskchanky commented on .rustfmt.toml:
```diff
@@ -0,0 +1,13 @@
+use_small_heuristics = "Default"
+force_explicit_abi = true
```
`true` is the default.
https://github.com/rust-lang/rustfmt/blob/master/Configurations.md#force_explicit_abi

raskchanky commented on .rustfmt.toml:
```diff
@@ -0,0 +1,13 @@
+use_small_heuristics = "Default"
+force_explicit_abi = true
+hard_tabs = false
+max_width = 100
```
100 is the default.
https://github.com/rust-lang/rustfmt/blob/master/Configurations.md#max_width

raskchanky commented on .rustfmt.toml:
```diff
@@ -0,0 +1,13 @@
+force_explicit_abi = true
+hard_tabs = false
+max_width = 100
+merge_derives = true
```
`true` is the default
https://github.com/rust-lang/rustfmt/blob/master/Configurations.md#merge_derives

baumanj replied:
I think I would prefer this one to be `false`. It's easy to manually merge the derives, but there are occasions where it's nice to keep them separate for clarity or debugging purposes.

raskchanky replied:
Fair point.  I'll change it.

raskchanky commented on .rustfmt.toml:
```diff
@@ -0,0 +1,13 @@
+merge_derives = true
+newline_style = "Auto"
+remove_nested_parens = true
+reorder_imports = true
```
`true` is the default
https://github.com/rust-lang/rustfmt/blob/master/Configurations.md#reorder_imports

raskchanky commented on .rustfmt.toml:
```diff
@@ -0,0 +1,13 @@
+reorder_imports = true
+reorder_modules = true
+tab_spaces = 4
+use_field_init_shorthand = true
```
`false` is the default for this one, which I changed to `true`, based on our conversations in the Clippy PR.
https://github.com/rust-lang/rustfmt/blob/master/Configurations.md#use_field_init_shorthand

raskchanky commented on .rustfmt.toml:
```diff
@@ -0,0 +1,13 @@
+reorder_modules = true
+tab_spaces = 4
+use_field_init_shorthand = true
+use_try_shorthand = true
```
`false` is the default for this one, which I changed to `true`, because that's what all of our code looks like.
https://github.com/rust-lang/rustfmt/blob/master/Configurations.md#use_try_shorthand

baumanj replied:
[Based on the docs](http://doc.rust-lang.org/1.32.0/core/macro.try.html), I'm not sure that the `?` operator is 100% equivalent to the `try!` macro, but I think it's good to migrate to the preferred form, and if there's a difference, I imagine we'll find out at compile time.

raskchanky commented on .rustfmt.toml:
```diff
@@ -0,0 +1,13 @@
+tab_spaces = 4
+use_field_init_shorthand = true
+use_try_shorthand = true
+edition = "2018"
```
2015 is the default for this one, but I changed it to 2018, because we've already gone to the effort of converting the repo over to be 2018 compliant.
https://github.com/rust-lang/rustfmt/blob/master/Configurations.md#edition

raskchanky commented on .rustfmt.toml:
```diff
@@ -0,0 +1,24 @@
+edition = "2018"
+
+# unstable options (for one-off fixes and CI)
+merge_imports = true
```
The default is `false` but I changed this to `true` because I think we've all agreed that this is a style we want to move towards and having all of that work done for us for free is amazing.
https://github.com/rust-lang/rustfmt/blob/master/Configurations.md#merge_imports

raskchanky commented on .rustfmt.toml:
```diff
@@ -0,0 +1,24 @@
+
+# unstable options (for one-off fixes and CI)
+merge_imports = true
+wrap_comments = true
```
The default is `false` but I changed it to `true` because [the issue that spawned this PR](https://github.com/habitat-sh/habitat/issues/6130) was about wrapping comments.
https://github.com/rust-lang/rustfmt/blob/master/Configurations.md#wrap_comments

raskchanky commented on .rustfmt.toml:
```diff
@@ -0,0 +1,24 @@
+# unstable options (for one-off fixes and CI)
+merge_imports = true
+wrap_comments = true
+comment_width = 80
```
`80` is the default.
https://github.com/rust-lang/rustfmt/blob/master/Configurations.md#comment_width

baumanj replied:
Any reason not to have it at 100 to match the code width? If people find 80 more readable, that's fine with me, but I like to save the vertical space that 100 provides.

raskchanky replied:
Ah, good call.  I spaced out on that.  No reason - I will change to 100.

raskchanky commented on .rustfmt.toml:
```diff
@@ -0,0 +1,24 @@
+merge_imports = true
+wrap_comments = true
+comment_width = 80
+condense_wildcard_suffixes = true
```
The default is `false` but I changed it to `true` because the code samples look a lot cleaner with this enabled, and it feels like less visual noise.
https://github.com/rust-lang/rustfmt/blob/master/Configurations.md#condense_wildcard_suffixes

raskchanky commented on .rustfmt.toml:
```diff
@@ -0,0 +1,24 @@
+wrap_comments = true
+comment_width = 80
+condense_wildcard_suffixes = true
+format_strings = true
```
The default is `false` but I changed this to `true`, thinking that it would lead to more compact, easier to read code.  I don't have strong feelings on this one way or the other though.
https://github.com/rust-lang/rustfmt/blob/master/Configurations.md#format_strings

baumanj replied:
I like having everything wrapped at the same column, so I'm in favor of this.

raskchanky commented on .rustfmt.toml:
```diff
@@ -0,0 +1,24 @@
+comment_width = 80
+condense_wildcard_suffixes = true
+format_strings = true
+normalize_comments = true
```
The default is `false` but I changed it to `true` because I like the consistency of having all comments look the same.  I'm not sure of other editors, but in my editor, if I select a block of code and use the keystroke to comment it, it inserts `//` at the beginning of every line, instead of using the `/* */` variant.  My hunch is that other editors work similarly, though I'll admit total ignorance on this point.  I also don't feel strongly about this one way or the other.
https://github.com/rust-lang/rustfmt/blob/master/Configurations.md#normalize_comments

raskchanky commented on .rustfmt.toml:
```diff
@@ -0,0 +1,24 @@
+condense_wildcard_suffixes = true
+format_strings = true
+normalize_comments = true
+reorder_impl_items = true
```
The default is `false` but I changed it to `true` because it feels like more organized code.
https://github.com/rust-lang/rustfmt/blob/master/Configurations.md#reorder_impl_items

raskchanky commented on .rustfmt.toml:
```diff
@@ -0,0 +1,24 @@
+format_strings = true
+normalize_comments = true
+reorder_impl_items = true
+format_doc_comments = true
```
The default is `false` but I changed it to `true` because why not have formatting inside doc comments?  😁 
https://github.com/rust-lang/rustfmt/blob/master/Configurations.md#format_doc_comments

raskchanky commented:
I've added the `DO-NOT-MERGE` label to this until @scotthain sorts out the BK stuff he's doing, and we can ditch travis and appveyor on this PR.  There are CI bits still to add to this PR, which I'm going to wait to tinker with until Scott is done.  After that's added, I'll format everything.  I had originally included all the formatting in the PR, but I've removed it because rebasing is too painful.  Once the CI bits are ready, I'll rebase, format everything, and merge.

raskchanky commented:
This is in its final form.  @baumanj @christophermaier @scotthain Anything else I should add/change?

baumanj commented on support/ci/lint.sh:
```diff
@@ -1,364 +0,0 @@
-#!/usr/bin/env bash
```
Yasssssss!

baumanj commented on support/ci/rustfmt.sh:
```diff
@@ -0,0 +1,23 @@
+
+# This is being pinned to a specific date because sometimes nightly will make a change that breaks rustfmt's usage of rustc.
+# We should decide how/when/if we want to advance this date pin.
+toolchain="nightly-2019-02-11"
```
I suggest we add this as [a release task](https://github.com/habitat-sh/habitat/blob/master/.github/ISSUE_TEMPLATE/release-checklist.md) to update this. Alternatively, we could make this use the most currently nightly and retry with a few slightly older nightlies dynamically if there's a failure.

raskchanky replied:
I like your second suggestion better, as it would ensure that CI is always able run `rustfmt`, vs having a human have to muck around figuring out what the last nightly was that included `rustfmt`.  I'll file an issue to handle this.

baumanj commented on components/hab/src/command/origin/key/upload.rs:
```diff
@@ -64,14 +68,7 @@ pub fn start(
-                 origin key. \
-                 Giving up.\
-                 ",
+                "We tried {} times but could not upload {}/{} public origin key. Giving up.",
```
Yassssss!
